### PR TITLE
Compaction in partition threads

### PR DESCRIPTION
**Pull request by ahmetmircik**
EE: https://github.com/hazelcast/hazelcast-enterprise/pull/5908

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/Storage.java [hunk 1]
```diff
@@ -36,6 +37,10 @@
  */
 public interface Storage<K, R> {
 
+    default Step newCompactorStep() {
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/Storage.java:40]: done.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepSupplier.java [hunk 2]
```diff
@@ -140,6 +140,8 @@ private void runStepWithState(Step step, State state) {
         boolean metWithPreconditions = true;
         try {
             try {
+                state.getRecordStore().beforeOperation();
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepSupplier.java]: done.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java [hunk 3]
```diff
@@ -130,6 +132,18 @@ public final void beforeRun() throws Exception {
...
+        tieredStoreEnabled = recordStore != null
+                && getStartingStep() != null
+                && mapConfig.getTieredStoreConfig().isEnabled();
+
+        if (mapStoreOffloadEnabled) {
+            assert !tieredStoreEnabled;
+        }
+
+        if (tieredStoreEnabled) {
+            assert !mapStoreOffloadEnabled;
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java]: put all assertions in a method: `assertOnlyOneOfMapStoreOrTieredStoreEnabled`

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java [hunk 4]
```diff
@@ -130,6 +132,18 @@ public final void beforeRun() throws Exception {
                 && getStartingStep() != null
                 && !mapConfig.getTieredStoreConfig().isEnabled();
 
+        tieredStoreEnabled = recordStore != null
+                && getStartingStep() != null
+                && mapConfig.getTieredStoreConfig().isEnabled();
+
+        if (mapStoreOffloadEnabled) {
+            assert !tieredStoreEnabled;
```
[the remark this answers is not in the document]
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java]: put all assertions in a method: `assertOnlyOneOfMapStoreOrTieredStoreEnabled`

**Comment by ahmetmircik** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/ClearOpSteps.java [hunk 5]
```diff
@@ -100,7 +100,7 @@ public void runStep(State state) {
 
         @Override
         public Step nextStep(State state) {
-            return UtilSteps.SEND_RESPONSE;
+            return UtilSteps.FINAL_STEP;
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/ClearOpSteps.java:103]: Is it just a renaming of constant or we change the semantics here?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/IMapOpStep.java [hunk 6]
```diff
@@ -87,6 +87,7 @@ default boolean isStoreStep() {
      * @return name of the executor to execute this map operation step
      */
     default String getExecutorName(State state) {
+        // TODO use separate pool for tstore enabled map
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/IMapOpStep.java:90]: Do you plan to use extra pool in this PR?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/UtilSteps.java [hunk 7]
```diff
@@ -25,6 +25,19 @@
 
 public enum UtilSteps implements IMapOpStep {
 
+    FINAL_STEP() {
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/UtilSteps.java:28]: So, for all tiered store operations, we will return this FINAL_STEP checking if the compaction is needed?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java [hunk 8]
```diff
@@ -140,7 +140,7 @@ public void runStep(State state) {
         @Override
         public Step nextStep(State state) {
             if (state.isStopExecution()) {
-                return UtilSteps.SEND_RESPONSE;
+                return UtilSteps.FINAL_STEP;
```
Comment (petrpleshachkov) [inline hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/PutOpSteps.java:143]: Is it correct that now for tiered store we always execute Put/GetOperation using Put/GetOpSteps mechanics? If do, where is it enabled?

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/UtilSteps.java (hunk 7 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/UtilSteps.java:36]: we will use injected step, final is a dispatcher step.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/IMapOpStep.java (hunk 6 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/IMapOpStep.java]: yes it is in EE PR now, please see CompactorStep in there.

File: hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/ClearOpSteps.java (hunk 5 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/ClearOpSteps.java:103]: no semantic change for other flows, final step is used to dispatch flow to injected steps.

**Comment by ahmetmircik** [comment]
moved to https://github.com/hazelcast/hazelcast/pull/24280, for a cleanup on my side.

**Pull request closed without merging**